### DISCUSSION TITLE
Maintainable design for multiple backend targets

### DISCUSSION THREAD

**Discussion by jeff-hykin** [vlang/v#13459] [General]
This is for preventing code duplication between the backend, and is a prerequisite to enhancing the browser backend with major new features.

**Comment by jeff-hykin** [comment]
Just as a disclaimer, currently I do not have any big picture perspective on the compilation architecture of V. I barely have a small picture. But I'm a programming language enthusiast, so I've been thinking about the design of systems and syntax for a long time.

I believe in "ideal" implementations of things, although compilation steps seem particularly tricky. I can't say I have a full design for an ideal multi-compile-targets system.

However, I'm confident there are few characteristics needed for an elegant/maintainable solution.

**I think there needs to be split between "surface", and "subterranean" code. **


A split where everything 'above' the surface only uses a kind of V lang extended library/api. If a standard library is something supported on all backends, the the extended library would a list of features supported on at least one backend (and ideally supported on all backends).

The "surface" itself would be part of V lang. It would be the part that imports unstable methods from the "subterranean" portion while exporting a stable external facing API. In my thought it would be pure V code, and contain logic like fallbacks,  polyfills, and most of all abstractions. 

The "subterranean" portion would be the messy V-code-mixed-with-js or mixed-with-C stuff. The current idea would be to have a large  flat list of files. The idea being that each file represents a tiny isolated backend feature, and would ideally be be little more than V lang function definition acting as a wrapper. This would result in an extremely large list of features. From here, defining a backend (like "browserjs") is simple; it's merely the a list of subterranean features that it supports. Targets like ES6, ES7, Deno, and node JS have many overlapping features but also many inherent incompatiblites. If someone spends the time, targets could even become more specific such as targeting Chrome with browser Bluetooth support or webGpu support. In theory a flat list like this would allow for maximum dry code without getting the C-macros kind of "dry" code where everything is just a mess of ifdef's. In my thought, no subterranean file would contain something like "if (NodeJS)", each isolated file would assume the lower level tool exists, and the "surface" code  handles the if (NodeJS) kind of logic and statements.


I do see two challenges though:
- interactions between features, for example where one feature needs a C #include, and another feature has already included it (and we don't wan to #include it twice). There might be a way to handle this elegantly with the surface code calling a subterranean feature, and passing it a `state` variable that it can use.
- compile time optimizations: I'm not sure how those are implemented in V lang

**Comment by JalonSolov** [comment]
Currently, in a very simplified description, the main V code is converted to an Abstract Syntax Tree (AST), then the backends generate code from there.

V supports separation of backend code by filename convention, with names matching `*.<backend>.v`, so C specific code should go in `*.c.v` files, JavaScript in `*.js.v` files, etc.  It also supports target OS specific code with names matching `*_<target>.v`, for example `*_windows.v` for Windows specific code, etc.  Now combine those for C code specifically for Windows, that code would go in `*_windows.c.v` files.

Also currently, there is a lot of C/JS specific code in regular .v files, but that will change over time, until it becomes a compiler error to put backend specific code in a plain/pure .v file.

That said, we can already split a lot of things about to specific files based on which backend is the target.  The good thing about this system is that V will ignore files (again, based on the name) that are not pure .v files and do not match the target OS/backend.  That part already works.

As for optimizations, V doesn't currently do a lot of optimizations at all, but what it does do is handled by the `vlib/v/transformer` code.  At the moment, if you truly want optimized code, compile with the `-prod` option and let the C compiler do the optimizations for you.

**Reply by jeff-hykin** [reply]
Thanks for that! It helps a lot to understand the methodology. 

I've dug around in some of the `.js.v` files so your posts help put those into context.

**Comment by jeff-hykin** [comment]
In that case I suppose the initial effort then should be to get all the C/JavaScript code into `.c.v` `.js.v` files.


So I'm guessing currently combinations of target outputs would involve treating them as an OS and then enumerating all possible combinations. Like:
- es5_node
- es5_broswer_no_webgpu
- es5_broswer_webgpu
- es6_node
- es6_broswer_no_webgpu
- es6_broswer_webgpu

**Reply by JalonSolov** [reply]
Well... all of those are JavaScript types, and right now, that's handled a little differently.  Run `v help build-js` at the command line, and you'll see the options there.  Those could be expanded, or we may need something like `_es5_node.js.v`, but that starts to get rather unwieldy.
